### DISCUSSION TITLE
error using ffplay

### DISCUSSION THREAD

**Discussion by Diyari90** [fraunhoferhhi/vvenc#288] [Q&A] [answered]
q1. error using ffplay for vvc file i used y4m file and convert it to h266 now it wont work by ffplay but when i do local stream it work dont know how i do correctly.
1.ffmpeg -i 1080.y4m -c:v vvc -b:v 2600k -preset faster out.h266 it work 
2. ffplay out.h266 not work
3. A. server ----- [ ffmpeg -i out.h266 -f mpegts udp://127.0.0.1:1234 ] work
B. clinet ------ [ ffplay udp://127.0.0.1:1234 ] work

q2. when i stream local i dont know how to record my video does anyone know which i write to doing it 

![Screenshot from 2023-06-24 02-32-23](https://github.com/fraunhoferhhi/vvenc/assets/33030503/09211bff-67e4-46b3-9d42-5ff226b73abe)
![Screenshot from 2023-06-24 03-04-39](https://github.com/fraunhoferhhi/vvenc/assets/33030503/15a829b2-c310-4388-ba45-ef1cd41ef347)

**Comment by adamjw24** [comment]
What FFmpeg are you using? It is not what we explain in [our Wiki](https://github.com/fraunhoferhhi/vvenc/wiki/FFmpeg-Integration). Please check with the person developing your FFmpeg integration why it wouldn't work.

**Reply by Diyari90** [reply]
thank i look at ffplay file inside codind it include library SDL it just need installin that  library some time it make misinstallingg ffplay,

but till now i dont now how to save local stream when i stream vvc video please if anyone know how to write that line for saveing stream local tell me thanks again for replay.

i  wirte those line the server work but clinet not saveing file it just creat crushing file with 7megabyte

**Comment by lehmann-c** [comment]
The ffplay version that your are using is the system installation that does not support vvc (version 4.4.2-ubuntu0.22.04.1 as can be seen in your screenshot).
Please check you PATH and/or that you install the ffmpeg with vvc support properly.
it´s the correct version, when the configuration contains: '--enable-libvvdec --enable-libvvenc"

**Reply by Diyari90** [reply]
thank i look at ffplay file inside codind it include library SDL it just need installin that  library some time it make misinstallingg ffplay,

but till now i dont now how to save local stream when i stream vvc video please if anyone know how to write that line for saveing stream local tell me thanks again for replay.

i  wirte those line the server work but clinet not saveing file it just creat crushing file with 7megabyte

**Reply by lehmann-c** [reply]
When you using the server as you mentioned above:
`ffmpeg -i out.h266 -f mpegts udp://127.0.0.1:1234`
then you are not streaming vvc, but mpeg2video (as you can also see in your screenshot)
So what happens is that ffmpeg transcodes the vvc input file to mpeg2 and streams that.

As far as I know transcoding vvc to ts or even streaming vvc does not work in ffmpeg.

So usually you have to use:
`ffmpeg -i out.h266 -vcodec copy -f mpegts udp://127.0.0.1:1234`

When you want to record a stream you have to store it with ffmpeg as well:
`ffmpeg -i udp://127.0.0.1:1234 -vcodec copy output.ts`

**Reply by Diyari90** [reply]
Thanks you right , i will try those line you wrote, iam new in this so probably i do mistake ❤️ 🌸🌷🌹🌺

**Reply by Diyari90** [reply]
@lehmann-c hi could you help me with saving while streaming
server:  --->  ffmpeg -re -i out.h266 -c copy -f vvc udp://127.0.0.1:1234
Client:  -----> ffplay udp://127.0.0.1:1234

both this line work good the quality not be bad while playing but if i use other line the stream is be bad quality
ffmpeg -i out.h266 -vcodec copy -f mpegts udp://127.0.0.1:1234
ffmpeg -i udp://127.0.0.1:1234 -vcodec copy output.ts

mpegt is bad quality for vvc but first 2 line is good but i don't know how to save it ,, what should i write to saving it

**Reply by Diyari90** [reply]
look when i want to saving my video its 500 frame i want get in client don't know why.
i used two time both time save error in client frame want be received 

![Screenshot from 2023-07-05 01-19-22](https://github.com/fraunhoferhhi/vvenc/assets/33030503/76a996fd-a30c-4778-b171-7e5c62f5abfa)
![Screenshot from 2023-07-05 01-24-51](https://github.com/fraunhoferhhi/vvenc/assets/33030503/70fadfa3-3467-453d-87ab-547d26aec969)

**Reply by lehmann-c** [reply]
This discussion is pretty off topic to vvenc.
What you are trying to do (streaming raw bitstreams without re-encoding) does not work with h264 or h265 as well in ffmpeg.

Nevertheless I´ll try to catchup with your problem:
1. Please do not copy screenshots, but the complete log from your terminal output, as all the main information is missing.  
2. The screenshots are still showing me that your are streaming mpeg2video (2nd screenshot: libvvdec -> mpeg2video).  
So you still are transcoding it somehow, but the screenshots are not matching to to commands that you have written above:
 `ffmpeg -i out.h266 -vcodec copy -f mpegts udp://127.0.0.1:1234`
  If you have bad quality, the video has been transcoded, and as you didn´t set any target bitrates, ffmpeg re-encodes it with 200kbit/s per default (which is pretty low)
4. and finally: As I mentioned before, remultiplexing vvc bitstreams into ts does not work in ffmpeg

If you want to stream a vvc bitstream as TS you should encode it as an ts (via ffmpeg) in the first place.
UPD streaming without re-encoding is a special topic that is not handled here. Please consider to read the ffmpeg guide for that.  

As a fist starting point I would  use tcp for testing:  
`ffmpeg -i input.ts -c:a copy -c:v copy -f mpegts  tcp://127.0.0.1:11000`  
`ffmpeg -i "tcp://localhost:11000?listen"  -vcodec copy -f mpegts output.ts`  

If that is working you could try to get udp working by using the ffmpeg guide.

**Reply by Diyari90** [reply]
thanks i see what you say, let me tell you my goal is to test vvc video how good and bad will streaming let just for second forget all i wrote if i want local stream video with vvc what should i write 1. in sever and 2. in clinet??for streaming and saving in client
i did install vvenc and vvdec and munally install ffmepg like it been written in this page (https://github.com/fraunhoferhhi/vvenc/wiki/FFmpeg-Integration)

**Reply by lehmann-c** [reply, marked as answer]
Just test it as I wrote above:

1. encode your input and generate a transport stream, e.g.:
`ffmpeg -i <input> -c:v vvc -b:v 2M -preset faster vvc.ts`

2. stream it via tcp client:
`ffmpeg -i vvc.ts -c:v copy -f mpegts  tcp://127.0.0.1:11000`

3. receive it by using an tcp server
`ffmpeg -i "tcp://localhost:11000?listen"  -vcodec copy -f mpegts output.ts`

If that works you can try udp streaming.

**Reply by Diyari90** [reply]
thanks but i want use udp not tcp i want to put on packet loss and delay to see it's effect, could you tell me what i should write for using udp for vvc video steaming

**Reply by JonnyTech** [reply]
Try replacing both instances of `tcp://` with `udp://`

**Reply by Diyari90** [reply]
Thanks

**Comment by Diyari90** [comment]
@lehmann-c Hi, i have i question :: how could i use rtp instead udp or tcp we mention it above.

i tried this but i make error
# Server
ffmpeg -re -i input.mp4 -c:v copy -f vvc -sdp_file test_video.sdp "rtp://127.0.0.1:1234"
# Client 
ffmpeg -protocol_whitelist rtp,udp -i "rtp://127.0.0.1:1234" out_stream.mkv

SDP not recognized--
-->[ No output streams in the SDP.
Error writing the SDP.
[vost#0:0/copy @ 0x557ae82f5e40] Error initializing output stream:  ]

**Reply by JonnyTech** [reply]
You are using a `SDP` file? Transfer `test_video.sdp` to the client - it is needed to receive the stream:

```
ffmpeg -protocol_whitelist "file,rtp,udp" -i test_video.sdp out_stream.mkv
```

**Reply by Diyari90** [reply]
I want to stream vvc video with rtp but I don’t know which command line i use, if you know could you write it for me

**Reply by JonnyTech** [reply]
No, I would just use UDP, did you use the SDP file correctly for RTP?

**Reply by Diyari90** [reply]
No i dont know how to us it correctly so i asked
